gensim/models/poincare.py

```diff
@@ -0,0 +1,827 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Author: Jayant Jain <jayantjain1992@gmail.com>
+# Copyright (C) 2017 Radim Rehurek <me@radimrehurek.com>
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+
+"""
+Python implementation of Poincare Embeddings, an embedding to capture hierarchical information,
+described in [1]
+
+This module allows training a Poincare Embedding from a training file containing relations from
+a transitive closure.
+
+
+.. [1] https://arxiv.org/pdf/1705.08039.pdf
+
+"""
+
+
+import csv
+import itertools
+import logging
+import os
+import random
+import time
+
+from autograd import numpy as np, grad
+from collections import defaultdict, Counter
+from numpy import random as np_random
+from smart_open import smart_open
+
+from gensim import utils
+from gensim.models.keyedvectors import KeyedVectors, Vocab
+from gensim.models.word2vec import Word2Vec
+
+logger = logging.getLogger(__name__)
+
+
+class PoincareModel(utils.SaveLoad):
```

Comment:
It's unclear what the public API is in this class. Looks like all the methods are private. It would actually make sense to prefix them with an underscore.

Then again, once the embedding is trained, it's not clear to me from the documentation (or even the code) what can I do with it.

Reply:
Yeah, it's only the `__init__` and `train` for now. I'd like to prefix the private methods too, just going with the existing gensim convention of not doing so. I don't mind changing if @menshikh-iv is okay with it.

We definitely need a clearer explanation of what can be done with the model post-training, for that, I need to resolve and finish the `KeyedVector` operations.

Reply:
@jayantj I agree with @janpom, prefixes is better, because I see the sad situation in global: with `existing gensim convention of not doing so` new user have many methods and doesn't know what should be called. Unfortunately, we can't fix this in the current code, because this change has broken all, but we can prevent this in new code.

Reply:
OK, let's prefix all the methods other than `train()` with underscore.

We could decide the same about the fields. I think it makes sense to keep most fields public since the user may want to read the values. @jayantj Please use your judgement there. Private should be the default. We don't want to clutter the API. On the other hand, we do want to keep things public where it makes sense.

Reply:
Done https://github.com/RaRe-Technologies/gensim/pull/1696/commits/530146dd91a859c5c4ccb75547e46f27c7540e0b and https://github.com/RaRe-Technologies/gensim/pull/1696/commits/be0249a42008dfbc6294fad9c246495e3c56fad6

I've kept most properties public. But I'm not sure what you mean by `private should be the default`, since the language itself doesn't have any notion of public/private (AFAIK)
Do you mean simply that we ourselves should mark any new method/property we add as public only if there's a specific reason to do so, and continue to mark them with an underscore if not?

Reply:
Looks good for me now, thanks!

@piskvorky wdyt about `_` and https://github.com/RaRe-Technologies/gensim/pull/1696#discussion_r149880020 ?